### PR TITLE
Mage flow pipeline

### LINKED ISSUE #14270: Mage-Flow
### Model/Pipeline/Scheduler description

Series of image generation models

Note: official implementation is hardcoded to CUDA and flash-attn 2 and may not be easy to port

### Open source status

- [x] The model implementation is available.
- [x] The model weights are available (Only relevant if addition is not a scheduler).

### Provide useful links for the implementation

https://github.com/microsoft/Mage/tree/main/mage_flow

https://huggingface.co/microsoft/Mage-Flow
https://huggingface.co/microsoft/Mage-Flow-Base
https://huggingface.co/microsoft/Mage-Flow-Turbo
https://huggingface.co/microsoft/Mage-Flow-Edit
https://huggingface.co/microsoft/Mage-Flow-Edit-Base
https://huggingface.co/microsoft/Mage-Flow-Edit-Turbo

### PR DESCRIPTION
**Pull request by chenyangzhu1**
# What does this PR do?

Add [Mage-Flow](https://github.com/microsoft/Mage) text-to-image pipeline to diffusers. Mage-Flow is a 4B parameter model from Microsoft's Mage team, based on the NR-MMDiT (Native-Resolution Multimodal DiT) dual-stream architecture with Qwen3-VL text encoder and a custom MageVAE (128-channel latents, 16x downsampling). Supports any resolution from 512 to 2048 at arbitrary aspect ratios.

This PR adds:
- `MageFlowTransformer2DModel` — dual-stream MMDiT with complex RoPE and joint attention
- `AutoencoderMageVAE` — DConv encoder + CoD decoder with adaLN constant folding
- `MageFlowPipeline` — T2I pipeline with Qwen3-VL text encoding and FlowMatchEuler scheduling
- Weight conversion script, tests (70 passed / 0 failed), and API documentation

## Usage

```python
import torch
from diffusers import MageFlowPipeline, MageFlowTransformer2DModel, FlowMatchEulerDiscreteScheduler
from diffusers.models.autoencoders.autoencoder_mage_vae import AutoencoderMageVAE
from transformers import Qwen3VLForConditionalGeneration, AutoTokenizer
from safetensors.torch import load_file

model_path = "path/to/converted/model"

scheduler = FlowMatchEulerDiscreteScheduler.from_pretrained(f"{model_path}/scheduler")
transformer = MageFlowTransformer2DModel(
    in_channels=128, out_channels=128, context_in_dim=2560,
    hidden_size=3072, num_attention_heads=24, num_layers=12,
    axes_dim=[16, 56, 56], patch_size=1,
)
transformer.load_state_dict(load_file(f"{model_path}/transformer/diffusion_pytorch_model.safetensors"))
transformer = transformer.to(torch.bfloat16).cuda()

vae = AutoencoderMageVAE()
vae.load_state_dict(load_file(f"{model_path}/vae/diffusion_pytorch_model.safetensors"))
vae = vae.to(torch.bfloat16).cuda()
vae.freeze_adaln()

text_encoder = Qwen3VLForConditionalGeneration.from_pretrained(
    f"{model_path}/text_encoder", torch_dtype=torch.bfloat16
).cuda()
tokenizer = AutoTokenizer.from_pretrained(f"{model_path}/text_encoder")

pipe = MageFlowPipeline(
    scheduler=scheduler, vae=vae, text_encoder=text_encoder,
    tokenizer=tokenizer, transformer=transformer,
)

image = pipe(
    prompt="A cat sitting on a windowsill, sunlight streaming in",
    height=1024, width=1024,
    num_inference_steps=30,
    guidance_scale=5.0,
).images[0]
image.save("mage_flow_output.png")
```

## Quality Verification

All images 1024x1024, 30 steps, CFG=5.0. The original pipeline uses Gaussian-Shading watermark noise and packed varlen text encoding, so exact pixel-level reproduction is not expected — but visual quality and prompt adherence are equivalent.

### Mountain Landscape (seed=88)

| Original Pipeline | Diffusers Pipeline |
|:---:|:---:|
| orig_mountain | diff_mountain |

### Cat on Windowsill (seed=123)

| Original Pipeline | Diffusers Pipeline |
|:---:|:---:|
| orig_cat | diff_cat |

### City Skyline (seed=77)

| Original Pipeline | Diffusers Pipeline |
|:---:|:---:|
| orig_city | diff_city |

### Numerical Parity

| Metric | Value |
|--------|-------|
| Transformer forward max_diff (bf16, GPU) | **0.000000** |
| Transformer forward cos_sim (bf16, GPU) | **0.99999994** |
| Weight key match (transformer) | 397 / 397 |
| Weight key match (VAE) | 728 / 728 |

> **Note on image differences**: Images are not pixel-identical because (1) the original pipeline uses Gaussian-Shading watermark noise while diffusers uses standard `randn`, (2) the original uses packed varlen text encoding while diffusers uses standard padded-batch encoding. Both produce equivalent visual quality.

### Transformer (`MageFlowTransformer2DModel`)
- NR-MMDiT dual-stream architecture with 12 dual-stream blocks
- Complex RoPE with symmetric positive/negative frequency scaling, stored as real-valued registered buffers for correct `model.to(device)` behavior
- Joint attention via `dispatch_attention_fn` (non-packed [txt, img] concatenation)
- Custom sinusoidal timestep embedding that downcasts the frequency table to input dtype (bf16) before computing — the model was trained with this exact rounding, using float32 frequencies produces visibly degraded output
- `latent_height`/`latent_width` forward params to avoid GPU-CPU sync in RoPE (torch.compile compatible)

### VAE (`AutoencoderMageVAE`)
- DConv encoder (one-step diffusion) + CoD decoder with DConv denoiser
- 128-channel latents, 16x spatial downsampling
- adaLN constant folding via `freeze_adaln()` for inference
- `F.scaled_dot_product_attention` in CoD decoder attention blocks
- `forward()` delegates to `decode()` for CPU offload hook compatibility

### Pipeline (`MageFlowPipeline`)
- Qwen3-VL (`Qwen3VLForConditionalGeneration`) text encoder with ChatML prompt template
- `FlowMatchEulerDiscreteScheduler(shift=6.0)` with custom base sigmas `linspace(1, 1/N, N)`
- Standard two-pass CFG with graceful fallback when text_encoder unavailable
- `model_cpu_offload_seq = "text_encoder->transformer->vae"`

### Weight Conversion
- `scripts/convert_mage_flow_to_diffusers.py`
- Transformer: `img_in` → `x_embedder`, `txt_norm` → `context_embedder_norm`, `txt_in` → `context_embedder`
- VAE: `student.dconv_encoder.*` → `encoder.*`, `pipeline.*` → `decoder.*` (excluding `y_embedder.encoder.*`)

## Test Results

```
73 passed, 0 failed, 3 skipped (torch 2.7.1+cu126)
```

The 3 skipped tests:
- 1x `test_save_load_optional_components`: Pipeline has no optional components (correct skip)
- 2x `test_keep_in_fp32_modules`: Mage-Flow does not need `_keep_in_fp32_modules` (correct skip)

## Self-Review

Ran `.ai/skills/self-review` against `.ai/review-rules.md`, `.ai/models.md`, `.ai/pipelines.md`, `.ai/testing.md`, and `.ai/skills/model-integration/pitfalls.md`.

### Blocking issues found and resolved

| # | Issue | Resolution |
|---|-------|------------|
| B1 | Transformer declared `encoder_hidden_states_mask` param but never used it | Removed from signature and pipeline calls |
| B2 | Transformer declared unused `txt_ids` param | Removed |
| B3 | Pipeline inlined custom sigma schedule | Added comment explaining why it differs from scheduler default |
| B4 | Dead `temb + torch.zeros(...)` no-op | Removed |
| B5 | `.item()` calls in RoPE break torch.compile | Added `latent_height`/`latent_width` params |
| B6 | Hardcoded developer paths in conversion script | Replaced with placeholders |
| B7 | Pipeline docstring referenced `AutoencoderKL` | Fixed to `AutoencoderMageVAE` |
| B8 | Tests imported via internal module paths | Changed to public API |

### Non-blocking issues found and resolved

| # | Issue | Resolution |
|---|-------|------------|
| NB1 | PosEmbed stored complex tensors as plain attributes | Split into real/imag registered buffers |
| NB2 | `_MageVAEYEmbedder` wrapper undocumented | Added docstring for weight-key compatibility |
| NB3 | Denoiser dispatched on channel count | Changed to explicit `is_latent` parameter |
| NB4 | VAE missing `_no_split_modules` | Added class attribute |
| NB5 | `encode()` lacked generator for reproducibility | Added `generator` param with `randn_tensor` |
| NB6 | Inline CFG | Consistent with Flux/QwenImage, kept as-is |
| NB9 | `MageVAEAttnBlock` used `torch.bmm` | Replaced with `F.scaled_dot_product_attention` |

### Dead code removed

| Location | What | Why dead |
|----------|------|----------|
| Transformer | `temb + zeros` block | No-op (adding zero) |
| Transformer | `scale_rope=False` branch | `scale_rope` always `True` |
| Transformer | `encoder_hidden_states_mask` / `txt_ids` params | Never read |
| VAE | `MageVAEDecoder.ada = nn.Identity()` | No-op identity layer |
| VAE | Duplicate `forward()` method | Superseded by decode-delegating forward |

### New Files
- `src/diffusers/models/transformers/transformer_mage_flow.py`
- `src/diffusers/models/autoencoders/autoencoder_mage_vae.py`
- `src/diffusers/pipelines/mage_flow/pipeline_mage_flow.py`
- `src/diffusers/pipelines/mage_flow/pipeline_output.py`
- `src/diffusers/pipelines/mage_flow/__init__.py`
- `scripts/convert_mage_flow_to_diffusers.py`
- `tests/pipelines/mage_flow/test_pipeline_mage_flow.py`
- `tests/models/transformers/test_models_transformer_mage_flow.py`
- `tests/pipelines/mage_flow/__init__.py`
- `docs/source/en/api/pipelines/mage_flow.md`
- `docs/source/en/api/models/mage_flow_transformer2d.md`

### Modified Files
- `src/diffusers/__init__.py` — Register new components
- `src/diffusers/models/__init__.py` — Register models
- `src/diffusers/models/transformers/__init__.py`
- `src/diffusers/models/autoencoders/__init__.py`
- `src/diffusers/pipelines/__init__.py` — Register pipeline
- `docs/source/en/_toctree.yml` — Add doc entries

Fixes https://github.com/huggingface/diffusers/issues/14270

## Before submitting
- [x] Did you use an AI agent (Claude Code, Codex, Cursor, etc.) to help with this PR? If so:
  - [x] Did you read the [Coding with AI agents](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution#coding-with-ai-agents) guide?
  - [x] Did you run the [`self-review`](https://github.com/huggingface/diffusers/blob/main/.ai/skills/self-review/SKILL.md) skill on the diff?
  - [x] Did you share the final self-review notes in the PR description or a comment?
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution)?
- [x] Did you read our [philosophy doc](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy)? (important for complex PRs)
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?
- [ ] Are you the author (or part of the team) of the model/pipeline (only applicable for model/pipeline related PRs)?

## Who can review?

@sayakpaul @yiyixuxu @DN6 @dg845 @asomoza

### CODE REVIEW

**Comment by Xinjie-Q** [comment]
We have reviewed the PR and confirmed that the implementation is correct. The Mage-Flow integration looks good, including the transformer, VAE, pipeline, weight conversion, tests. From our side, the PR is ready and can be merged into Diffusers. Thanks for the contribution!

**Comment by alerikaisattera** [comment]
Is edit pipeline included?

**Comment by yiyixuxu** [comment]
thanks so much for the PR @chenyangzhu1 !

Could you convert the pipeline to Modular Diffusers?

We are currently trying to do modular-only releases whenever possible — as long as we're not under release time pressure, we prefer new integrations to be released in modular pipeline from the start (it should be a pretty straightforward task for your AI agent if you're using one! )

File: src/diffusers/__init__.py [hunk 1]
```diff
@@ -697,6 +699,8 @@
             "LTXLatentUpsamplePipeline",
             "LTXPipeline",
             "LucyEditPipeline",
+            "MageFlowPipeline",
```
Comment (alvarobartt) [inline src/diffusers/__init__.py:702]: Note that you should also include this in https://github.com/alvarobartt/diffusers/blob/a8c1c08f42f321ca853f8f4349ba3ede0ca4358b/src/diffusers/pipelines/auto_pipeline.py#L138-L191 so that the `AutoPipelineForText2Image` can be leveraged for this model; it's really just about adding this entry:

```python
        ("mage-flow", MageFlowPipeline),
```

cc @asomoza for visibility
